123-job-jwt/proposal.md

```diff
@@ -0,0 +1,137 @@
+a static credential to Concourse. This alleviates the need for a human to handle
+sensitive credentials, increases security by ensuring credentials are not
+statically configured and JWTs are ephemeral to each run, and reduces
+Concourse's need to properly handle and encrypt data (can be limited to storing
+the JWT private signing key).
+
+## Proposal
+
+Adopt [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519) and a Go library
+that [implements](https://github.com/golang-jwt/jwt) it. Additionally adopt
+[RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517) and a Go library that
+implements it. (Perhaps [jwx](https://github.com/lestrrat-go/jwx)?)
+
+Create a struct that includes as many unique job metadatum as possible,
+including standard claims of `iss`, `sub`, `aud`, `exp`, `jti`, `nbf`, `iat`,
+and Concourse standard objects such as `team`, `teamid`, `job`, `jobid`,
+`pipeline`, `pipelineid`, `task`, `buildid`, and any other data points that
+describe the job and could drive authentication and authorization claims.
+
+Create a public API endpoint that responds to requests where the signed JWT is
+part of the payload, likely leveraging JWKS.
+
+## Open Questions
+
+- Where should the claims be aggregated and signing portion be completed? (atc?)
+- How should the signed JWT be delivered to the container runtime?
+- Where should the validation endpoint live? (web?)
+- Where should the signed JWT live? (Environment Variable + File?)
+
+## Answered Questions
+
+(This section intentionally left blank)
+
+## New Implications
+
+Since it is a new feature, placed into optional paths, it should not
+significantly impact existing workflows. A potential issue may arise if the
+pipeline defines resources or executes job tasks that interact with the proposed
+locations for the signed JWT in the environment. However, if the placement of
+the JWT happens early enough in the lifecycle, this can be mitigated as the
+environment variable would simply be overwritten by the user config. A file
+location would present a harder collision, but there should be reserved paths
+for Concourse artifacts, potentially even mounted to the container as a
+read-only filesystem.
+
+Because we are proposing that the JWT be used externally with products such as
+Hashicorp Vault, the JWT would do nothing unless configured to be trusted by the
+external system. Thus, the impact to existing pipelines would be minimal.
+
+## Appendixes
+
+### Potential Workflow in Hashicorp Vault context
+
+A Hashicorp Vault operator would have to mount and configure the JWT Auth Method
+like so:
+
+```sh
+curl -X POST -H "X-Vault-Request: true" -H "X-Vault-Token: $(vault print token)" -d '{"type":"jwt","description":"","config":{"options":null,"default_lease_ttl":"0s","max_lease_ttl":"0s","force_no_cache":false},"local":false,"seal_wrap":false,"external_entropy_access":false,"options":null}' https://127.0.0.1:8200/v1/sys/mounts/concourse
```

Comment:
I'm not super familiar with vault, but is this all config that needs to be done inside Vault to allow JWTs from concourse to access that namespace? I'd assume we'd need to point vault to some `.well-known/openid-configuration` endpoint, so it would know where to periodically fetch the JWKs from, and possibly describe other restrictions about which tokens are allowed to access that namespace…

Reply:
You are right. Vault needs one of three things for jwt-auth:
a) statically configured keys (bad)
b) a link to a JWKS with the (better)
c) a link to an OIDC-Configuration endpoint

There are also a few other things that may be configured here. For example you can configure vault to expect a certain value in the aud-claim of the token.

Most importantly a policy must be specified that is to be used for identities that were created by jwt-auth.
The good thing is that values from the token can be used inside the policy. It would therefore for example be easy to write a policy that grants access to all the secret belonging to the pipeline.